### DISCUSSION TITLE
Cant free my provider because of reference count

### DISCUSSION THREAD

**Discussion by AbadiNiro** [openssl/openssl#26243] [Q&A]
Hi,
My provider custom teardown function don't get called ,
I can see that the refcnt is 2 and openssl wont invoke the custom teardown ,

the order of creation and deletion is as follows:
OSSL_LIB_CTX_new //create an empty OSSL_LIB_CTX
OSSL_PROVIDER_add_builtin // added my custom provider
OSSL_PROVIDER_load // load the default provider into my OSSL_LIB_CTX which created earlier
OSSL_PROVIDER_load// load the custom provider into my OSSL_LIB_CTX which created earlier

in the deletion
OSSL_PROVIDER_unload // unload the custom provider into my OSSL_LIB_CTX which created earlier
OSSL_PROVIDER_unload// unload the default provider into my OSSL_LIB_CTX which created earlier
OSSL_LIB_CTX_free // release my LIB_CTX

i cant understand why my refcnt of the custom provider while unloading is 2
is it the right order of deletion?

**Comment by beldmit** [comment]
`test/provfetchtest.c` is probably the example similar to what you are doing, could you please check what happens with refcount there?

**Reply by AbadiNiro** [reply]
Thanks for your response , I will check it.
I can see that in the example the teardown function is set to be OSSL_LIB_CTX_free 
and there is call for OSSL_PROVIDER_unload only in case of an error.

so should i also call only for OSSL_LIB_CTX_free ?

**Reply by beldmit** [reply]
> so should i also call only for OSSL_LIB_CTX_free ?

I don't think so but I didn't read the teardown documentation.

**Reply by AbadiNiro** [reply]
Even if I unload the provider right after the load function the refcnt is 2 ,
because in the load function its been created (refcnt is 1) and than added to store (refcnt is 2)
so what is the right behavior in order to get to provider actually freed

**Reply by mattcaswell** [reply]
Did you fetch any algorithms from you provider? Have all fetched algorithms been freed?

**Reply by AbadiNiro** [reply]
Nothing explicitly
The things I created explicitly
OSSL_LIB_CTX_new
OSSL_PROVIDER_load
OSSL_PROVIDER_load
SSL_CTX_new_ex
SSL_new
EVP_PKEY_CTX
EVP_PKEY_fromdata

and all of them are released bottom up

(inside the OSSL_LIB_CTX_free function I can see that the default teardown called but the custom teardown isnt after the refcnt is 1 its returned)

**Reply by mattcaswell** [reply]
> and all of them are released bottom up

So, IIUC, you free the EVP_PKEY, EVP_PKEY_CTX, SSL and SSL_CTX first, before unloading the providers and freeing the OSSL_LIB_CTX?

Hmm....this should be sufficient to free all ref counts AFAICT.

**Reply by AbadiNiro** [reply]
yes and I can see that all their refcnt is zero and theyre actually freed

**Reply by mattcaswell** [reply]
Can you simplify your code to see which call is taking the ref count and not releasing it?

**Reply by mattcaswell** [reply]
> because in the load function its been created (refcnt is 1) and than added to store (refcnt is 2)

`OSSL_PROVIDER_unload` should decrement 1 refcnt, and `OSSL_LIB_CTX_free` should decrement the other.

**Reply by AbadiNiro** [reply]
```
OSSL_LIB_CTX libctx=OSSL_LIB_CTX_new();
OSSL_PROVIDER_load(libctx,"default");
OSSL_PROVIDER_load(libctx,"custom");
SSL_CTX sslctx = SSL_CTX_new_ex(libctx,"?provider=default" ,TLS_client_method());
SSL ssl = SSL_new(sslctx);
EVP_PKEY_CTX evpctx = EVP_PKEY_CTX_new_from_name(libctx,keyname,"provider=custom");
EVP_PKEY_fromdata(evpctx,&pkey,...)
SSL_use_certificate
SSL_use_privateKey
```

my custom provider implements the signature and keymgmt operations 
I can see that every resource allocated getting freed properly besides my custom provider

if any additional information needed please let me know

**Reply by mattcaswell** [reply]
If you remove all the `SSL_CTX_*`, `SSL_*, and `EVP_PKEY_*` calls from your code (i.e. just leaving creating/freeing the libctx and loading/unloading the providers - does the problem go away?

**Reply by AbadiNiro** [reply]
Yes, if Im unloading after loading, the teardown function get called properly

**Reply by mattcaswell** [reply]
What if you just add back the EVP_PKEY_CTX and EVP_PKEY_fromdata calls?

**Reply by AbadiNiro** [reply]
I add back everything beside SSL_use_PrivateKey
and its working , so this is the problematic function but i cant understand why

**Reply by mattcaswell** [reply]
Do you free the `SSL` and the `EVP_PKEY`?

**Reply by AbadiNiro** [reply]
Yes, inside of this function I can see that my OSSL_FUNC_KEYMGMT_NEW got invoked and I dont see any free call

**Reply by mattcaswell** [reply]
Can you run this in a debugger? It would be helpful to understand what is happening to the "references" field of the `EVP_PKEY` structure. You will need to compile OpenSSL with debug symbols for that.

**Reply by AbadiNiro** [reply]
Yes im on it , will update

**Reply by AbadiNiro** [reply]
The SSL_use_PrivateKey increment it to 2 as expected 
and it get freed properly, the refcnt of the provider stay as is

**Reply by mattcaswell** [reply]
Do I understand correctly that the references drop to 0 as expected for the EVP_PKEY, but you don't see a subsequent OSSL_FUNC_KEYMGMT_FREE call as a result? This is expected to happen immediately.

You might set a breakpoint in EVP_PKEY_free and watch what happens when the ref count drops to 0.

https://github.com/openssl/openssl/blob/b85e6f534906f0bf9114386d227e481d2336a0ff/crypto/evp/p_lib.c#L1770-L1809

EVP_PKEY_free() should call the internal function `evp_pkey_free_it` if references drops to 0. The `evp_pkey_free_it` function should then call `evp_keymgmt_freedata` which should call your OSSL_FUNC_KEYMGMT_FREE function.

You might also watch what is happening to the `refcnt` field in the `OSSL_PROVIDER` object for your provider while all of this is happening.

**Reply by AbadiNiro** [reply]
I can see that it is happening and that side works fine,
I also noticed that if im calling SSL_use_PrivateKey without calling SSL_use_certificate the teardown also got called,

so if one of these functions dont get called the teardown get called , if they both got called the teardown dont get called

**Reply by mattcaswell** [reply]
So what is the OSSL_PROVIDER `refcnt` value before you call `SSL_use_privateKey` or `SSL_use_certificate`? And what is it after you have called each of those functions? And again what is it after you have called `EVP_PKEY_free` and `SSL_free`?

**Reply by mattcaswell** [reply]
I assume you do also free the certificate (X509_free)?
